server/text_generation_server/models/causal_lm.py

```diff
@@ -230,6 +226,33 @@ def recombine(cls, batches: List["CausalLMBatch"], is_optimized_for_gaudi: bool
         for b in batches:
             b.past_key_values = list(b.past_key_values)
 
+        # For prefill there is a space allocated only for first token
+        # Need to add padding to the max total tokens before first decode
+        for batch in batches:
+            padding_needed = (batch.input_length + batch.right_padding) - batch.seq_length
+            if padding_needed > 0:
+                batch.input_ids = torch.nn.functional.pad(batch.input_ids, (0, batch.right_padding), value=pad_token_id)
+                batch.attention_mask = torch.nn.functional.pad(batch.attention_mask, (0, batch.right_padding), value=0)
+
+                past_key_values = []
+                value_padding = (0, 0, 0, batch.right_padding)
+                if key_dim == -1:
+                    key_padding = (0, batch.right_padding)
+                else:
+                    key_padding = (0, 0, 0, batch.right_padding)
+                for layer_num in range(len(batch.past_key_values)):
+                    updated_key = torch.nn.functional.pad(batch.past_key_values[layer_num][0], key_padding, value=0)
+                    updated_value = torch.nn.functional.pad(batch.past_key_values[layer_num][1], value_padding, value=0)
+                    past_key_values.append((updated_key, updated_value))
+                    batch.past_key_values[layer_num] = None
+                batch.past_key_values = past_key_values
+                del past_key_values
+                htorch.core.mark_step()
+
+        max_seq_len = batches[0].seq_length
+        input_length = max_input_length
+        right_padding = max_seq_len - input_length
+
```

Comment:
Could we break down this block into smaller chunks and move them below so that everything that touches input_ids is in the same place, everything that touches attention_mask is in another place and so on?

Reply:
I've updated PR, please check